src/controllers/auth.controller.js

```diff
@@ -0,0 +1,277 @@
+  }
+
+  const userData = jwtService.verify(token);
+
+  if (!userData) {
+    throw ApiError.badRequest('Invalid or expired token');
+  }
+
+  const user = await User.findByPk(userData.id);
+
+  if (!user) {
+    return next(ApiError.badRequest('User not found'));
+  }
+
+  const hashedPass = await bcrypt.hash(password, 10);
+
+  user.password = hashedPass;
+
+  await user.save();
+
+  res.send({ message: 'Password successfully reset' });
+};
+
+const updateProfile = async (req, res, next) => {
+  const {
+    name,
+    oldPassword,
+    newPassword,
+    confirmPassword,
+    newEmail,
+    confirmEmail,
+  } = req.body;
+  const userId = req.user.id;
+
+  const user = await User.findByPk(userId);
+
+  if (!user) {
+    return next(ApiError.notFound('User not found'));
+  }
+
+  if (name) {
+    if (name.trim().length < 2) {
+      return next(
+        ApiError.badRequest('Name should be at least 2 characters long'),
+      );
+    }
+
+    user.name = name.trim();
+  }
+
+  if (newPassword) {
+    if (!oldPassword) {
+      return next(ApiError.badRequest('Old password is required'));
+    }
+
+    if (newPassword !== confirmPassword) {
+      return next(
+        ApiError.badRequest('New password and confirmation do not match'),
+      );
+    }
+
+    const isValidPassword = await bcrypt.compare(oldPassword, user.password);
+
+    if (!isValidPassword) {
+      return next(ApiError.badRequest('Invalid old password'));
+    }
+
+    const hashedPassword = await bcrypt.hash(newPassword, 10);
+
+    user.password = hashedPassword;
+  }
+
+  if (newEmail) {
+    if (newEmail !== confirmEmail) {
+      return next(
+        ApiError.badRequest('New email and confirmation do not match'),
+      );
+    }
+
+    if (!oldPassword) {
+      return next(
+        ApiError.badRequest('Old password is required for email change'),
+      );
+    }
+
+    const isValidPassword = await bcrypt.compare(oldPassword, user.password);
+
+    if (!isValidPassword) {
+      return next(ApiError.badRequest('Invalid old password for email change'));
+    }
+
+    const emailError = validateEmail(newEmail);
+
+    if (emailError) {
+      return next(ApiError.badRequest('Invalid email', { email: emailError }));
+    }
+
+    const existingUser = await userService.findByEmail(newEmail);
+
+    if (existingUser) {
+      return next(ApiError.badRequest('Email already in use'));
+    }
+
+    const oldEmail = user.email;
+
+    await sendConfirmationEmail(oldEmail);
+    user.email = newEmail;
+  }
+
+  await user.save();
+  res.send(userService.normalize(user));
+};
+
+const generateToken = async (res, user) => {
+  const normalizedUser = userService.normalize(user);
+
+  const accessToken = jwtService.sign(normalizedUser);
+  const refreshAccessToken = jwtService.signRefresh(normalizedUser);
+
+  await tokenService.save(normalizedUser.id, refreshAccessToken);
+
+  res.cookie('refreshAccessToken', refreshAccessToken, {
+    maxAge: 30 * 24 * 60 * 1000,
```

Comment:
Consider setting a more secure expiration time for the refresh token. The current setting of 30 days might be too long for some applications.